File: App/Resources/AlgorithmDetails/treesort/treesort.swift
final class Node {
    let pointer: Int
    var left: Node?
    var right: Node?

    init(_ pointer: Int) {
        self.pointer = pointer
    }
}

func sort(_ arr: inout [Int]) {
    let n = arr.count
    var root: Node?

    func add(_ node: Node?, _ addPtr: Int) -> Node {
        guard let node = node else {
            return Node(addPtr)
        }
        if arr[addPtr] < arr[node.pointer] {
            node.left = add(node.left, addPtr)
        } else {
            node.right = add(node.right, addPtr)
        }
        return node
    }

    for i in 0 ..< n {
        root = add(root, i)
    }

    var result: [Int] = []

    func traverse(_ node: Node?) {
        guard let node = node else {
            return
        }
        traverse(node.left)
        result.append(arr[node.pointer])
        traverse(node.right)
    }

    traverse(root)

    for i in 0 ..< n {
        arr[i] = result[i]
    }
}

var array: [Int] = [
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
]
sort(&array)
print(array)

File: App/Resources/AlgorithmDetails/treesort/treesort.py
class Node:
    def __init__(self, pointer):
        self.pointer = pointer
        self.left = None
        self.right = None


def sort(arr):
    n = len(arr)
    root = None

    def add(node, add_ptr):
        if node is None:
            return Node(add_ptr)
        if arr[add_ptr] < arr[node.pointer]:
            node.left = add(node.left, add_ptr)
        else:
            node.right = add(node.right, add_ptr)
        return node

    for i in range(n):
        root = add(root, i)

    result = []

    def traverse(node):
        if node is None:
            return
        traverse(node.left)
        result.append(arr[node.pointer])
        traverse(node.right)

    traverse(root)

    for i in range(n):
        arr[i] = result[i]
    return arr


if __name__ == "__main__":
    array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56]
    sort(array)
    print(array)
